src/queue.c

```diff
@@ -951,6 +997,9 @@ libdispatch_init(void)
 	dispatch_assert(sizeof(struct dispatch_root_queue_context_s) %
 			DISPATCH_CACHELINE_SIZE == 0);
 
+#if DISPATCH_USE_THREAD_LOCAL_STORAGE
+	_dispatch_thread_key_create(&__dispatch_tsd_key, (void(*)(void*))_libdispatch_thread_cleanup);
```

Comment:
your function takes `(void)` which is why you need the cast and that is not the right way to do it.

`_libdispatch_thread_cleanup` should be:

```C
static void
_libdispatch_tsd_cleanup(void *ctxt DISPATCH_UNUSED)
{
...
```

And actually I think the context shouldn't be "unused", even if you "know" its address, it's cleaner to cast the `ctxt` into your struct type and clean this instead of reloading the __thread thing.